### DISCUSSION TITLE
Weird behaviour with dynamic route - initial state doesn't include query

### DISCUSSION THREAD

**Discussion by remy** [vercel/next.js#15952] [Help] [answered]
I've got a simple page that uses `useRouter` and I'm seeing the router query come in _after_ page render:

```js
import { useRouter } from 'next/router';

export default function Page() {
  const router = useRouter();
  const { courseId } = router.query;
  console.log(courseId); // once as undefined, then again with actual value
  
  return <h1>{ courseId }</h1>
}
```

Surely this should have the actual query populate on SSR? 

I've also tried wrapping in `withRouter` - same effect. Also the log on the server side is always undefined.

**Comment by remy** [comment]
Oh, is it due to this one single footnote on this page? https://nextjs.org/docs/routing/dynamic-routes

**Reply by Timer** [reply]
Correct!
https://nextjs.org/docs/routing/dynamic-routes#caveats

> Pages that are statically optimized by Automatic Static Optimization will be hydrated without their route parameters provided, i.e query will be an empty object ({}).

The query cannot be provided at hydration time otherwise you could have a hydration mismatch.

In your case (I see "courses"), I'd suggest using `getStaticProps` and `getStaticPath` that'd let you prerender all of your courses statically.

**Reply by vixeven** [reply]
@Timer What is the workaround for this scenario: a browse page with filters, which can be optionally (it means query object can be `{}` as URL haven't any parameters, just like the default `hydration`).

For this page, filters have to be populated from `query`. Also, how to avoid a potentially useless request?

Importing everything dynamically with `{ ssr: false }` is the only solution I've got so far.

**Comment by remy** [comment, marked as answer]
Because I don't see how you can sensibly use SWR along side dynamic routes (or certainly I've not come across any docs) and it seems like a pretty obvious use case: https://swr.vercel.app/#overview

**Reply by Timer** [reply]
It's pretty easy! SWR supports [conditional data fetching](https://swr.vercel.app/docs/conditional-fetching), which is how we use SWR on [vercel.com](https://vercel.com).

```tsx
import { useRouter } from "next/router";

export default function Page() {
  const router = useRouter();
  const { courseId } = router.query;
  const { data } = useSWR(courseId ? `/api/data?course=${courseId}` : null);

  return <h1>{courseId}</h1>;
}
```

**Comment by remy** [comment]
The data isn't static and based on users, so I have to use swr, which could really do with that example, where you've basically hacked around the problem by passing a null fetcher. 

My solution was to double nest the components - equally as hacky.

**Comment by ozumahe** [comment]
Some times the router query come in after page render: You can use React useEffect to  perform your action when ever the query changes

```
  useEffect(() => {
 
    if (router.query.categoryId) {
      getItem(router.query.categoryId);
    }
  }, [router.query.categoryId]);

```

**Reply by igortas** [reply]
Not working for useCallback calls insde useEffect.
